### PR TITLE
Abstracted complexity away from plugin initializer.

### PR DESCRIPTION
**Pull request by maumercado**
This PR is based on the init branch.

Abstracted specific Zipkin configuration to the plugin itself, the user should only set the service name, and Zipkin URL, however, a tracer is an optional argument the user can also send, whenever the need for a custom tracer is required, like in the tests.

Updated dependencies.

Remove test for Zipkin sending the correct http.url with query params, since support for url was removed for security reasons, and now in exchange for that there is http.path.

Updated tests accordingly to new Zipkin API.

cc @lucamaraschi & @mattcan

#### Checklist

- [ x] run `npm run test` and `npm run benchmark`
- [ x] tests and/or benchmarks are included
- [ x] documentation is changed or added
- [ x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -88,38 +84,6 @@ test('Should register the hooks and trace the request (404)', t => {
...
-
-      const annotations = record.args.map((args) => args[0])
-      t.strictEqual(annotations[2].annotation.annotationType, 'BinaryAnnotation')
-      t.strictEqual(annotations[2].annotation.key, 'http.url')
-      t.strictEqual(annotations[2].annotation.value, '/?foo=bar')
-
-      t.end()
-    })
-  })
-})
```
Comment (mcollina) [inline test.js]: Why was this test removed?

File: index.js [hunk 2]
```diff
@@ -3,19 +3,32 @@
...
 const None = zipkin.option.None
 const Instrumentation = zipkin.Instrumentation
 
 function zipkinPlugin (fastify, opts, next) {
-  const tracer = opts.tracer
-  if (tracer == null) return next(new Error('No tracer specified'))
+  assert(opts.serviceName, 'serviceName option should not be empty')
+  const { Tracer, BatchRecorder, jsonEncoder: { JSON_V2 } } = require('zipkin')
+  const { HttpLogger } = require('zipkin-transport-http')
+  const CLSContext = require('zipkin-context-cls')
```
Comment (mcollina) [inline index.js]: Can you move these requires to the top of the file?

File: package.json [hunk 3]
```diff
@@ -22,16 +22,17 @@
...
+    "node-fetch": "^2.6.0",
+    "sinon": "^7.5.0",
     "standard": "^10.0.3",
-    "tap": "^11.0.1",
-    "zipkin-context-cls": "^0.11.0",
-    "zipkin-transport-http": "^0.11.2"
+    "tap": "^14.10.2",
+    "zipkin": "^0.19.1",
+    "zipkin-context-cls": "^0.19.0",
+    "zipkin-transport-http": "^0.19.1"
```
Comment (mcollina) [inline package.json]: I think the zipkin dependencies should be dependencies, not devDependencies.

File: test.js (hunk 1 above)
Comment (maumercado) [reply test.js]: Removed that test, because, for security reasons, Zipkin has removed support for http.url and instead are only using http_path property, which does not contain the query parameters, and so that test becomes irrelevant with the new version of Zipkin.

File: index.js (hunk 2 above)
Comment (maumercado) [reply index.js]: Done! 👍

File: package.json (hunk 3 above)
Comment (maumercado) [reply package.json]: Done! 👍

File: test.js (hunk 1 above)
Comment (maumercado) [reply test.js]: https://github.com/openzipkin/zipkin-js/pull/226

File: .travis.yml [hunk 4]
```diff
@@ -1,10 +1,8 @@
 language: node_js
 
 node_js:
-  - "9"
-  - "8"
-  - "6"
-  - "4"
+  - "12"
+  - "10"
```
Comment (Eomm) [inline .travis.yml]: node 8 has been dropped by which dependancy?

File: index.js [hunk 5]
```diff
@@ -3,19 +3,32 @@
...
+const CLSContext = require('zipkin-context-cls')
 
 const Some = zipkin.option.Some
 const None = zipkin.option.None
 const Instrumentation = zipkin.Instrumentation
 
 function zipkinPlugin (fastify, opts, next) {
-  const tracer = opts.tracer
-  if (tracer == null) return next(new Error('No tracer specified'))
+  assert(opts.serviceName, 'serviceName option should not be empty')
```
Comment (Eomm) [inline index.js]: we should add a test for this condition

File: index.js [hunk 6]
```diff
@@ -3,19 +3,32 @@
...
+
+  const recorder = new BatchRecorder({
+    logger: new HttpLogger({
+      endpoint: `${opts.zipkinUrl}/api/v2/spans`,
```
Comment (Eomm) [inline index.js]: is not mandatory this parameter?

File: index.js [hunk 7]
```diff
@@ -3,19 +3,32 @@
...
+      jsonEncoder: JSON_V2
+    })
+  })
+
+  const ctxImpl = new CLSContext('zipkin')
+  const tracer = new Tracer({ ctxImpl, recorder, localServiceName: opts.serviceName })
```
Comment (Eomm) [inline index.js]: could we avoid to create it if `opts.tracer` has been provided?

File: .travis.yml (hunk 4 above)
Comment (maumercado) [reply .travis.yml]: huh... will add it back.

File: test.js [hunk 8]
```diff
@@ -8,17 +8,49 @@ const zipkin = require('zipkin')
...
+test('Should error when initializing the plugin without serviceName argument', t => {
+  const fastify = Fastify()
+
+  const record = sinon.spy()
+  const recorder = { record }
+  const ctxImpl = new ExplicitContext()
+  const tracer = new Tracer({ recorder, ctxImpl })
+  const zipkinUrl = 'http://0.0.0.0:9441'
+
+  ctxImpl.scoped(() => {
```
Comment (mcollina) [inline test.js]: Why is this needed here?

File: index.js [hunk 9]
```diff
@@ -66,6 +85,6 @@ function basic404 (req, reply) {
 }
 
 module.exports = fp(zipkinPlugin, {
-  fastify: '>=0.40.0',
+  fastify: '>= 2',
```
Comment (mcollina) [inline index.js]:
```suggestion
  fastify: '^2.0.0',
```

So that it fails on Fastify v3.

File: test.js (hunk 8 above)
Comment (maumercado) [reply test.js]: yeah we don't need that.. removing

File: index.js (hunk 6 above)
Comment (maumercado) [reply index.js]: It is, so, added a test for it, and an assertion in case a value is not set.
